blocks/bricks/__init__.py

```diff
@@ -158,7 +159,56 @@ def _push_initialization_config(self):
                     child.biases_init = self.biases_init
 
 
-class Linear(Initializable):
+class Feedforward(Brick):
+    """Declares an interface for bricks with one input and one output.
+
+    Many bricks have just one input and just one output (activations,
+    :class:`Linear`, :class:`MLP`). To make such bricks interchangable
+    in most contexts they should share an interface for configuring
+    their input and output dimensions. This brick provides such an
+    interface.
+
+    """
+    @abstractproperty
+    def input_dim():
+        """The input dimension of the brick."""
+        pass
+
+    @abstractproperty
+    def output_dim():
+        """The output dimension of the brick."""
+        pass
+
+
+def override_with_attribute(property_, attribute=None):
```

Comment:
An `abstractproperty` does not give an error on access! It gives an error when a class does not implement the property at all. But now you're forcing the user to explicitly implement a property (that mimics an attribute) if they just want to use an attribute (like in `Linear` where you just want to do `self.input_dim = input_dim`)... That's not very nice at all. A user should be able to just use an attribute if he wants to.

The only requirement I see is that at the time of allocation, there is an `input_dim` and `output_dim` that makes sense. This is the same requirement we have for every other attribute on every other brick. I think that in this case we just need to forgo the complexity of forcing a user to use a non-obvious way of overriding an abstract property with a property that acts like an attribute in order to set the attribute. Just reading that sentence makes me even more convinced that this is really adding needless complexity; just add to the docstring: "FeedForward bricks are expected to have a `input_dim` and `output_dim` attribute accessible on allocation."